walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,114 @@
+  }
+
+  /** 
+  * Creates a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //check if any people in meeting request is present in this set
+      boolean mtngAttendeeInEvent = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));
+
+      if (mtngAttendeeInEvent) {
+        // If there's already an event with this start time in the map, but the current end time is later, add the current instead.  
+        int eventStartTime = event.getWhen().start();
+        int eventEndTime = event.getWhen().end();
+        if (mandatoryEvents.containsKey(eventStartTime)) {
+          if (eventEndTime > mandatoryEvents.get(eventStartTime)) {
+            mandatoryEvents.put(eventStartTime, eventEndTime);
+          }
+        } else {
+          mandatoryEvents.put(eventStartTime, eventEndTime);
+        }
+      }
+    }
+    return mandatoryEvents;
+  }
+
+  /**
+  * Creates and returns an ArrayList of TimeRange objects that span available meeting times.
+  */
+  public Collection<TimeRange> findMeetingTimes (long mtngDuration, HashMap<Integer, Integer> mandatoryEvents) {
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    int currTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      // If there's an event starting at this time, see if there is a possible meeting time. 
+      // Reset the current duration variable, move the current time to the end time of the event, and repeat loop. 
+      if (mandatoryEvents.containsKey(currTime)) {
+        int currentLongest = mandatoryEvents.get(currTime);
+        if (currDuration >= mtngDuration) {
+          // Not inclusive because a meeting starts at this time 
+          meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, /* inclusive = */ false));
+        }
+        currDuration = 0;
+
+        // Check if there is a conflict: another meeting that starts during this one, and goes later
+        int longestMeetingEnd = getLongestEventEnd(currTime, currentLongest, mandatoryEvents);
+            
+        currTime = longestMeetingEnd;
+        startTime = longestMeetingEnd;
+      } else {
+        // If there is no event at this time, add MIN_INCREMENT minutes to the duration and current time, and run the loop again. 
+        currDuration += MIN_INCREMENT;
+        currTime += MIN_INCREMENT;
+      }
+    }
+
+    // When reaching the end of loop, if the current duration is longer than what is necessary, add it to the return Collection, inclusive. 
+    if (currTime > TimeRange.END_OF_DAY && currDuration >= mtngDuration) {
```

Comment:
When execution gets here, `currTime` could be (will be, assuming `END_OF_DAY` is a multiple of `MIN_INCREMENT`, I think) == `END_OF_DAY`. So shouldn't the first comparison be `>=`?

Reply:
yeah, should be